lib/elixir/lib/task.ex

```diff
@@ -338,13 +341,17 @@ defmodule Task do
   end
 
   @doc """
-  Receives a group of tasks and a message and finds
+  Takes a list of tasks and a message and finds
   a task that matches the given message.
 
-  This function returns a tuple with the returned value
-  in case the message matches a task that exited with
-  success alongside the matching task. It returns `nil`
-  if no task was found. It exits if the task has failed.
+  If the message matches a task that finished normally,
+  this function returns a tuple with the task's
+  result and `Task` struct.
+
+  If the message matches a task that terminated before
+  sending its result back, this function will exit.
```

Comment:
I agree with the two ways of using tasks you describe, those are the essential building blocks for performing async tasks concurrently.

Neither of them fits well into a gen server though. I'm not proposing to remove `Task.find` but I take it that you share my concerns about its utility. It almost seems to me that it should have behaved like `Task.yield` and returned `{:exit, reason}` instead of exiting. Then it would have been able to work nicely together with `async_nolink`. But let's just finish this PR and defer any further design discussion until another time.

My last question would then be: should we leave the `## Compatibility with OTP behaviours` section I added above? Perhaps, in the doc for `find/2` I should reiterate the point about adjusting the task code to return a tagged tuple from `async/3`'s doc:

> If you don't want the task to fail then you must change the `heavy_fun/0` code in the same way you would achieve it if you didn't have the async call. For example, to either return `{:ok, val} | :error` results or, in more extreme cases, by using `try/rescue`.

I don't @fishcakez would be happy about people sticking `try/rescue` in their tasks, but, in my opinion, this would serve as a workaround approximating the desired behaviour of `Task.find` I described in the second paragraph of this comment.

Regardless, I'm willing to just remove the `OTP Compatibility` section above and merge this if you don't have any further comments about my changes.